linera-core/src/value_cache.rs

```diff
@@ -153,16 +153,23 @@ impl CertificateValueCache {
         let hash = value.hash();
         let mut cache = self.cache.lock().await;
         if cache.contains(&hash) {
-            return false;
-        }
-        if let Some(confirmed_value) = value.validated_to_confirmed() {
-            // Cache the certificate for the confirmed block in advance, so that the clients don't
-            // have to send it.
-            cache.push(confirmed_value.hash(), confirmed_value);
+            // Promote the re-inserted value in the cache, as if it was accessed again.
+            cache.promote(&hash);
+            if let Some(confirmed_value) = value.validated_to_confirmed() {
+                // Also promote its respective confirmed certificate value
+                cache.promote(&confirmed_value.hash());
```

Comment:
Could it potentially happen that between line 157 and 160 enough time passes for the confirmed value to get evicted? What does `promote` do in that case?

(Or between 167 and 170—then the confirmed one would get evicted earlier.)

Reply:
Right, but lines 167 and 170 have the consequence that the confirmed value is _older_, so it gets evicted _first_.

Reply:
Oh, I see. I'll fix that, thanks! Sorry, I had misunderstood what you meant :sweat_smile:

Reply:
Done. The specific changes (before the rebase) can be seen [here](https://github.com/jvff/linera-protocol/compare/promote-reinserted-values-in-cache-r2..promote-reinserted-values-in-cache-r2-fixes).